File: Anurag_19_CSE/Java/day18.java
// Link For Problem: https: // leetcode.com/problems/critical-connections-in-a-network/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class day18 {

    /*
     * TC : O(V+E)
     * 
     * SC : O(V)
     */

    static int time = 0;

    public List<List<Integer>> criticalConnections(int n, List<List<Integer>> connections) {
        List<Integer> graph[] = new ArrayList[n];
        int disc[] = new int[n];

        Arrays.fill(disc, -1);
        int low[] = new int[n];

        List<List<Integer>> res = new ArrayList<List<Integer>>();
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<Integer>();
        }

        for (List<Integer> c : connections) {
            int src = c.get(0);
            int dest = c.get(1);
            graph[src].add(dest);
            graph[dest].add(src);
        }

        int parent[] = new int[n];
        Arrays.fill(parent, -1);

        for (int i = 0; i < n; i++) {
            if (disc[i] == -1)
                bridge(i, disc, low, parent, graph, res);
        }

        return res;
    }

    void bridge(int u, int disc[], int low[], int parent[], List<Integer> graph[], List<List<Integer>> res) {
        disc[u] = low[u] = time;
        time += 1;

        for (int v : graph[u]) {
            if (v == u)
                continue;

            if (disc[v] == -1) {
                parent[v] = u;
                bridge(v, disc, low, parent, graph, res);
                low[u] = Math.min(low[u], low[v]);

                if (low[v] > disc[u]) {
                    res.add(Arrays.asList(u, v));
                }
            } else if (v != parent[u]) {
                low[u] = Math.min(low[u], disc[v]);
            }
        }
    }
}

File: Anurag_19_CSE/Python/day18.py
# Link For Problem: https: // leetcode.com/problems/critical-connections-in-a-network/


class Solution:

    """ 
        TC : O(V+E)

        SC : O(V)
    """

    def criticalConnections(self, n: int, connections: list[list[int]]) -> list[list[int]]:

        # node is index, neighbors are in the list
        graph = [[] for i in range(n)]

        # build graph
        for n1, n2 in connections:
            graph[n1].append(n2)
            graph[n2].append(n1)

        # min_discovery_time of nodes at respective indices from start node
        # 1. default to max which is the depth of continuous graph
        lows = [n] * n

        # critical edges
        critical = []

        # args: node, node discovery_time in dfs, parent of this node
        def dfs(node, discovery_time, parent):

            # if the low is not yet discovered for this node
            if lows[node] == n:

                # 2. default it to the depth or discovery time of this node
                lows[node] = discovery_time

                # iterate over neighbors
                for neighbor in graph[node]:

                    # all neighbors except parent
                    if neighbor != parent:

                        expected_discovery_time_of_child = discovery_time + 1
                        actual_discovery_time_of_child = dfs(
                            neighbor, expected_discovery_time_of_child, node)

                        # nothing wrong - parent got what was expected => no back path
                        # this step is skipped if there is a back path
                        if actual_discovery_time_of_child >= expected_discovery_time_of_child:
                            critical.append((node, neighbor))

                        # low will be equal to discovery time of this node or discovery time of child
                        # whichever one is minm
                        # if its discovery time of child - then there is a backpath
                        lows[node] = min(
                            lows[node], actual_discovery_time_of_child)

            # return low of this node discovered previously or during this call
            return lows[node]

        dfs(n-1, 0, -1)

        return critical
